src/main/java/bio/terra/service/dataset/flight/UnlockDatasetStep.java

```diff
@@ -2,50 +2,61 @@
 
 import bio.terra.common.exception.RetryQueryException;
 import bio.terra.service.dataset.DatasetDao;
-import bio.terra.service.dataset.exception.DatasetLockException;
+import bio.terra.service.dataset.flight.ingest.OptionalStep;
 import bio.terra.stairway.FlightContext;
-import bio.terra.stairway.Step;
+import bio.terra.stairway.FlightMap;
 import bio.terra.stairway.StepResult;
 import bio.terra.stairway.StepStatus;
+import java.util.Objects;
 import java.util.UUID;
+import java.util.function.Predicate;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class UnlockDatasetStep implements Step {
+public class UnlockDatasetStep extends OptionalStep {
 
   private static Logger logger = LoggerFactory.getLogger(UnlockDatasetStep.class);
 
   private final DatasetDao datasetDao;
-  private UUID datasetId;
   private boolean sharedLock; // default to false
+  private UUID datasetId;
 
-  public UnlockDatasetStep(DatasetDao datasetDao, UUID datasetId, boolean sharedLock) {
+  public UnlockDatasetStep(
+      DatasetDao datasetDao,
+      UUID datasetId,
+      boolean sharedLock,
+      Predicate<FlightContext> doCondition) {
+    super(doCondition);
     this.datasetDao = datasetDao;
     this.datasetId = datasetId;
 
     // this will be set to true for a shared lock, false for an exclusive lock
     this.sharedLock = sharedLock;
   }
 
+  public UnlockDatasetStep(DatasetDao datasetDao, UUID datasetId, boolean sharedLock) {
+    this(datasetDao, datasetId, sharedLock, OptionalStep::alwaysDo);
+  }
+
   public UnlockDatasetStep(DatasetDao datasetDao, boolean sharedLock) {
-    this(datasetDao, null, sharedLock);
+    this(datasetDao, null, sharedLock, OptionalStep::alwaysDo);
   }
 
-  @Override
-  public StepResult doStep(FlightContext context) {
-    // In the create case, we won't have the dataset id at step creation. We'll expect it to be in
-    // the working map.
-    if (datasetId == null) {
-      datasetId = context.getWorkingMap().get(DatasetWorkingMapKeys.DATASET_ID, UUID.class);
-      if (datasetId == null) {
-        return new StepResult(
-            StepStatus.STEP_RESULT_FAILURE_FATAL,
-            new DatasetLockException("Expected dataset id in working map."));
-      }
-    }
+  public UnlockDatasetStep(
+      DatasetDao datasetDao, boolean sharedLock, Predicate<FlightContext> doCondition) {
+    this(datasetDao, null, sharedLock, doCondition);
+  }
 
+  @Override
+  public StepResult doOptionalStep(FlightContext context) {
     boolean rowUpdated;
+    FlightMap map = context.getWorkingMap();
     try {
+      // In the create case, we won't have the dataset id at step creation. We'll expect it to be in
+      // the working map.
+      datasetId =
```

Comment:
You need to catch the `NullPointerException` that results if the datasetId isn't in the working map and return a failed step result.

Reply:
Oh right, thanks!

Reply:
If the intent is to fail if `DatasetWorkingMapKeys.DATASET_ID` isn't in the working map, I think it would be clearer to avoid the NPE catch and check for the case explicitly.
```
    if (datasetId == null) {
      if (map.containsKey(DatasetWorkingMapKeys.DATASET_ID)) {
        datasetId = map.get(DatasetWorkingMapKeys.DATASET_ID, UUID.class);
      } else {
        return StepResult(
           StepStatus.STEP_RESULT_FAILURE_FATAL,
           new DatasetLockException(
               "Expected dataset id to either be passed in or in the working map."));
     }
   }
```

Reply:
Thanks for the discussion!